ipv8/attestation/trustchain/community.py

```diff
@@ -60,6 +59,9 @@ def __init__(self, *args, **kwargs):
         self.logger.debug("The trustchain community started with Public Key: %s",
                           self.my_peer.public_key.key_to_bin().encode("hex"))
         self.broadcast_block = True  # Whether we broadcast a full block after constructing it
+        self.listeners_map = {}  # Map of block_type -> [callbacks]
```

Comment:
Could you refactor the listeners into a single `Listener` class? The listener object, `should_sign` and `block_class` should all be connected anyway. This is starting to look like the TunnelCommunity.

Reply:
@qstokkink the `block_class` and `Listener` are not necessarily connected, they can be different objects.

Also, there can be multiple listeners for one block type but there's at most one `block_class` for each type. I don't see how we could generalize this nicely. Do you have any suggestions?

Reply:
I don't see how a listener could possibly comprehend what it is called back with, if it does not know the block class it is getting called back with. `self.notify_listeners(block)` gives a block of type `block = self.get_block_class(block_type)`.

A Listener could be something like:

```python
class TrustChainBlock(SomeBlockSuperclass):
    TYPE = 'bandwidth!'

class MyBeautifulListener(SomeListenerSuperclass):
    BLOCK_CLASS = TrustChainBlock

    def should_sign(block):
        return True
```